language/stdlib/modules/DesignatedDealer.move

```diff
@@ -144,10 +180,18 @@ module DesignatedDealer {
     spec fun add_tier {
         pragma opaque = true;
         modifies global<TierInfo<CoinType>>(dd_addr);
+        ensures exists<TierInfo<CoinType>>(dd_addr);
         ensures global<TierInfo<CoinType>>(dd_addr).window_start == old(global<TierInfo<CoinType>>(dd_addr)).window_start;
         ensures global<TierInfo<CoinType>>(dd_addr).window_inflow == old(global<TierInfo<CoinType>>(dd_addr)).window_inflow;
         ensures len(global<TierInfo<CoinType>>(dd_addr).tiers) == len(old(global<TierInfo<CoinType>>(dd_addr)).tiers) + 1;
         ensures global<TierInfo<CoinType>>(dd_addr).tiers[len(global<TierInfo<CoinType>>(dd_addr).tiers) - 1] == tier_upperbound;
+
+        include Roles::AbortsIfNotTreasuryCompliance{account: tc_account};
+        aborts_if !exists<TierInfo<CoinType>>(dd_addr);
+        let tier_info = global<TierInfo<CoinType>>(dd_addr);
+        let number_of_tiers = len(tier_info.tiers);
+        aborts_if number_of_tiers + 1 > MAX_NUM_TIERS with Errors::INVALID_ARGUMENT;
+        aborts_if number_of_tiers > 0 && tier_info.tiers[number_of_tiers - 1] >= tier_upperbound with Errors::INVALID_ARGUMENT;
```

Comment:
Same question here as in `update_tier` about proving that the other (existing) tiers remain unchanged

Reply:
I have made the spec of update_tier and add_tier more precise.